### PR TITLE
feat: Add HiCu model and MIMIC4ICD10Coding task for ICD-10 coding

### PR DESCRIPTION
**Pull request by matthew-ardi**
## Contributor

Matthew Ardi · NetID `mardi2` · CS 598 DL4H, Spring 2026

## Type

Model + Task (reuses existing `MIMIC4Dataset`).

## Paper

Ren et al., *HiCu: Leveraging Hierarchy for Curriculum Learning in Automated ICD Coding*, ML4H 2022 — https://arxiv.org/abs/2208.02301

## Summary

- `pyhealth.models.HiCu` — MultiResCNN encoder + per-label attention decoder with a 3-level ICD-10 hierarchy (chapter → 3-char category → full code). `set_depth(d)` activates depth `d` and copies parent weights into child positions so each curriculum stage starts warm. Uses Asymmetric Loss (Ben-Baruch et al. 2020).
- `pyhealth.tasks.MIMIC4ICD10Coding` — one sample per admission from discharge notes + `diagnoses_icd`. Filters to `icd_version == "10"` since MIMIC-IV mixes ICD-9 and ICD-10. Whitespace-tokenized, truncated to 4000 tokens.
- Curriculum training loop lives in the example (`model.set_depth()` between stages), so the model class stays stateless and works with the existing `Trainer`.

## Design decisions

- **ICD-10 on MIMIC-IV** instead of the paper's ICD-9 on MIMIC-III — current coding standard, newer dataset.
- **3 depths** (chapter / category / full) instead of 5. Covers the ICD-10 structure cleanly; extending to 5 is a follow-up.
- Label remapping uses a 1-D `scatter_add_` index buffer instead of a dense one-hot matrix (~6.6 MB → ~18 KB per depth on a 2,281-code vocab).
- Dropped Poincaré embeddings — paper reports marginal gains, and it keeps the PR dependency-free.

## Cerification
Verified on MIMIC-IV dev mode (1 000 patients / 907 samples / 2 281 codes) on an M2 Max with MPS

## Ablation (dev-mode, for path verification)

Schedule `5 / 10 / 30` epochs across depths.

| Config | Final train loss |
|---|---|
| Curriculum + ASL | 0.0007 |
| Flat + ASL | 0.0002 |
| Curriculum + BCE | 0.0020 |
| Curriculum + ASL (100 filters) | 0.0001 |

These are training losses on a small dev subset — absolute numbers aren't comparable to published benchmarks, and ASL depresses the scale by construction. The useful signal is the ~3× gap between ASL and BCE, which matches the paper's motivation. Proper benchmarking would need a held-out test split and micro-F1 / macro-F1 / P@k.

### CODE REVIEW

File: pyhealth/tasks/medical_coding.py [hunk 1]
```diff
@@ -139,44 +139,69 @@ def __call__(self, patient: Patient) -> List[Dict]:
...
+            tokens = _tokenize_clinical_text(raw_text)
+
+            if not tokens or len(icd_codes) < 1:
+                continue
+
+            samples.append(
+                {
+                    "patient_id": patient.patient_id,
+                    "text": tokens,
+                    "icd_codes": list(set(icd_codes)),
```
[the remark this answers is not in the document]
Comment (matthew-ardi) [reply pyhealth/tasks/medical_coding.py]: addressed

File: pyhealth/models/hicu.py [hunk 2]
```diff
@@ -0,0 +1,393 @@
...
+        embedded = self.word_embedding(text)
+        encoded = self.encoder(embedded.permute(0, 2, 1))
+        logits = self.decoder(encoded)
+
+        y_true_full = kwargs[self.label_key].to(self.device).float()
+        y_true = self._remap_labels(y_true_full, self.current_depth)
+        loss = self.asl_loss(logits, y_true)
+        y_prob = torch.sigmoid(logits)
+
+        return {"loss": loss, "y_prob": y_prob, "y_true": y_true, "logit": logits}
```
[the remark this answers is not in the document]
Comment (matthew-ardi) [reply pyhealth/models/hicu.py]: addressed

**Review by jhnwu3 — APPROVED** [review]
clever way of reusing the sequence processor here.

**Review by jhnwu3 — CHANGES_REQUESTED** [review]
commenting to come back to this if CIs succeed
